gateleen-core/src/main/java/org/swisspush/gateleen/core/future/SequentialFutures.java

```diff
@@ -0,0 +1,22 @@
+package org.swisspush.gateleen.core.future;
+
+import java.util.Iterator;
+import java.util.function.BiFunction;
+
+/**
+ * Executes Futures sequentially and stops when the first Future failed.
+ * Based on the solution mentioned here: https://stackoverflow.com/questions/51513678/reduce-future-sequentially-in-vert-x
+ *
+ * @author https://github.com/mcweba [Marc-Andre Weber]
+ */
+public class SequentialFutures {
+
+    public static <A, B> B execute(Iterator<A> iterator, B identity, BiFunction<B, A, B> function) {
+        B result = identity;
+        while(iterator.hasNext()) {
+            A next = iterator.next();
+            result = function.apply(result, next);
+        }
+        return result;
+    }
```

Comment:
This is basically a Stream.reduce(). Also, this code is actually not specific to futures.
Can we not replace this with:
```
list.stream().reduce(Future.succeededFuture(), (f, item) -> f.compose( action(item) ))
```
In the stackoverflow article, they did the combine (compose) after, which is the setup for parallelizing reduce. If we do it directly, it should work.

Reply:
You are right, I was able to replace the _SequentialFutures_ class with a "simple" _Stream.reduce()_